### PR TITLE
[8.0][IMP][mail_tracking] Track click events.

### PR DESCRIPTION
**Pull request by ivantodorovich**
Implements tracking click events on links, as discussed in https://github.com/OCA/social/issues/122

@antespi I took you suggestion, except for this one:

> Create a new table in order to save the relation of link to an unique token.

Instead, it replaces `href` values from anchor tags like this: `tracking_email_url/?url=original_href`
Then it routes those requests and redirects to the original URL, saving that event.
Inspired by how google does it for search result links

### CODE REVIEW

File: mail_tracking/models/mail_tracking_email.py [hunk 1]
```diff
@@ -170,23 +171,67 @@ def _compute_date(self):
...
+                    g.group(2)),
+                email['body']
+            )
+        return email
+
+    # Should I leave it as an alias? TODO: Remove later on
+    # @api.multi
+    # def tracking_img_add(self, email):
+    #     self.ensure_one()
+    #     return self._tracking_img_add(email)
```
Comment (ivantodorovich) [inline mail_tracking/models/mail_tracking_email.py]: Should I? Or should I not?

**Comment by ivantodorovich** [comment]
I'm wondering if there should be an on/off switch for this functionality.
It completely breaks links if the odoo instance is not exposed to the internet.

**Comment by ivantodorovich** [comment]
Ping @pedrobaeza @antespi

**Comment by antespi** [comment]
@ivantodorovich Thanks for your contribution. You should add a test case for click event.

File: mail_tracking/tests/test_mail_tracking.py [hunk 2]
```diff
@@ -126,6 +126,13 @@ def test_mail_send(self):
         self.assertEqual(image, res.response[0])
         # Two events again because no tracking_email_id found for False
         self.assertEqual(2, len(tracking.tracking_event_ids))
+        # Click event on links should redirect
+        http.request.httprequest.url = \
+            'http://www.odoo.com/?url=http://www.google.com.ar'
+        res = controller.mail_tracking_click(db, tracking.id)
+        self.assertEqual('301', res.status)
+        self.assertEqual('http://www.google.com.ar', res.headers['Location'])
+        self.assertEqual(3, len(tracking.tracking_event_ids))
```
Comment (ivantodorovich) [inline mail_tracking/tests/test_mail_tracking.py:135]: @antespi something like this is enough? I'm sorry I don't have much experience writing tests 😢

**Comment by ivantodorovich** [comment]
@antespi please check now and please, if its ok to you, remove the `work in progress` label so others can review.

**Comment by ivantodorovich** [comment]
After working with this changes in my local instance, I'm thinking there should be a way to enable/disable links tracking, because it's a very intrusive feature: it routes all links through our server, and that might not be what you want all the time.

I'm thinking of doing this by reading some attribute in the anchor element. For instance:

``` html
<a href="http://www.google.com" tracking-disabled>Click here</a>
```

or

``` html
<a href="http://www.google.com" tracking-enabled>Click here</a>
```

Then, it will only replace urls to route through our server if its required to do so.

**The question is: default enabled or default disabled?**
I'm more inclined to the latter, because of how intrusive this feature is.

What do you think?

**Comment by pedrobaeza** [comment]
I think a better strategy is the one taken in v9 in the link tracker. You create a link (like the shorteners goo.gl tc.o and similars), for your final link, and you paste in your mail this link, not the original one. Obviously, a way to do this from the mail itself would be more comfortable.

**Comment by ivantodorovich** [comment]
So, basically, no automatic link replacement?
User has to manually create the tracking link and replace it in his email?

I can live with that, but I think maybe offering a way to do it automatically depending on some customized attribute might be better for the user.

**Comment by pedrobaeza** [comment]
The best option would be to have a new mark in the link wizard to put "Track this link". In this case, you switch the provided link with this new link. In v8, directly the needed address for the controller with the corresponding token. In v9, creating a link in "Link tracker" and providing the result URL.

What do you think about this?

**Comment by ivantodorovich** [comment]
I'm not sure I understand your idea.

Let me explain what I understood to see if we are talking about the same thing.

**In v8:**
Track links that have `odoo-track-this-link` attribute, like this:

``` html
<a href="http://www.google.com" odoo-track-this-link>Click here</a>
```

Those links will be automatically replaced with the address of the controller that will track and redirect.

**In v9:**
Also track links that have `track-this-link` attribute. exactly like v8.

``` html
<a href="http://www.google.com" odoo-track-this-link>Click here</a>
```

But instead of defining a new controller, it will create a native 'tracking link' and use that url. 
It will also, somehow, catch the tracking event to be added in `mail_tracking`

`odoo-track-this-link` can be something else.

---

So, basically, this module will provide a faster way to generate tracking links (in v9) and also report click events as it does for mail open events.

**Comment by rafaelbn** [comment]
I change label to need review as all test are green @antespi

**Comment by rafaelbn** [comment]
@antespi @carlosdauden could you please take a review here? Thanks!
